lib/logstash/inputs/beats.rb

```diff
@@ -0,0 +1,215 @@
+        next if connection.nil? # if the connection is nil the connection was close.
+        connection = ConnectionDecorator.new(EventDecoration.new,
+                                             @codec.clone, connection)
+        invoke(connection) do |event|
+   	      if stop?
+            connection.close
+            break
+          end
+
+          begin
+            @circuit_breaker.execute {
+              @buffered_queue.push(event, @congestion_threshold)
+            }
+          rescue => e
+            raise e
+          end
+        end
+      else
+        @logger.warn("Beats input: the pipeline is blocked, temporary refusing new connection.")
+        sleep(RECONNECT_BACKOFF_SLEEP)
+      end
+    end
+  end # def run
+
+  public
+  def stop
+    @lumberjack.close
+  end
+
+  private
+  def invoke(connection, &block)
+    @threadpool.post do
+      begin
+        # If any errors occur in from the events the connection should be closed in the
+        # library ensure block and the exception will be handled here
+        connection.run(&block)
+
+        # When too many errors happen inside the circuit breaker it will throw
+        # this exception and start refusing connection. The bubbling of theses
+        # exceptions make sure that the lumberjack library will close the current
+        # connection which will force the client to reconnect and restransmit
+        # his payload.
+      rescue LogStash::CircuitBreaker::OpenBreaker,
+             LogStash::CircuitBreaker::HalfOpenBreaker => e
+        logger.warn("Beats input: The circuit breaker has detected a slowdown or stall in the pipeline, the input is closing the current connection and rejecting new connection until the pipeline recover.", :exception => e.class)
+      rescue => e # If we have a malformed packet we should handle that so the input doesn't crash completely.
+        @logger.error("Beats input: unhandled exception", :exception => e, :backtrace => e.backtrace)
+      end
+    end
+  end
+
+  def start_buffer_broker(output_queue)
+    @threadpool.post do
+      while !stop?
+        output_queue << @buffered_queue.pop_no_timeout
+      end
+    end
+  end
+
+  class EventDecoration < LogStash::Inputs::Base
+    public
+    def decorate(event)
+      super(event)
+    end
+  end # class EventDecoration
+
+  class ConnectionDecorator < SimpleDelegator
+    def initialize(event_decoration, codec, connection)
+      super(connection)
+      @event_decoration = event_decoration
+      @codec = codec
+    end
+
+    public
+    def run(&block)
+      super do |type, event|
+        case type
+        when :json; json_event(event, &block)
+        when :data; data_event(event, &block)
+        else; raise "unknown event type received"
+        end
+      end
+    end
+
+    private
+    def json_event(map, &block)
+      if map.is_a?(Array)
+        map.each { |item| json_event(item, &block) }
+      else
+        line = map.delete("line")
```

Comment:
Yes, that's right. in beats the 'message' field is renamed to "line" right now. Once the codec is configurable I would like to use 'line' in filebeat again as default line output (like LSF does). So elasticsearch will see 'line' and not 'message'.

With filebeat potentially writing to elasticsearch instead of logstash, file encoding might be dealt with in filebeat.